udpsession.go

```diff
@@ -0,0 +1,633 @@
+			}
+
+		case byte(pb.HeaderType_SESSION):
+			err := us.handleMetaData(buf[1:n], udpAddr)
+			if err != nil {
+				log.Printf("UdpSession.recvMsg handleMetaData error %v", err)
+			}
+
+		default:
+			log.Printf("UdpSession.recvMsg unknown format message")
+		}
+	}
+}
+
+func (us *UdpSession) Read(b []byte) (int, error) {
+	n, _, err := us.ReadFrom(b)
+	return n, err
+}
+
+// ReadFrom, implement of standard library net.UDPConn.WriteTo
+// Only for user data write
+func (us *UdpSession) WriteTo(b []byte, addr net.Addr) (n int, err error) {
+	buf := make([]byte, 1+len(b))
+	buf[0] = byte(pb.HeaderType_USER)
+	copy(buf[1:], b)
+
+	var udpAddr *net.UDPAddr
+	if us.isListener { // only listener need to use udpAddr
+		udpAddr, err = us.getUdpAddr(addr)
+		if err != nil {
+			return 0, err
+		}
+	}
+
+	conn, err := us.GetEstablishedUDPConn()
+	if err != nil {
+		return 0, err
+	}
+	n, _, err = conn.WriteMsgUDP(buf, nil, udpAddr)
+	if err != nil {
+		return 0, err
+	}
+
+	return n - 1, nil
+}
+
+func (us *UdpSession) Write(b []byte) (int, error) {
+	if us.isListener {
+		return 0, ErrOnlyDailer
+	}
+	return us.WriteTo(b, us.peer)
+}
+
+func (us *UdpSession) SetDeadline(t time.Time) error {
+	if us.ts.IsClosed() {
+		return ErrClosed
+	}
+
+	conn, err := us.GetEstablishedUDPConn()
+	if err != nil {
+		return err
+	}
+	return conn.SetDeadline(t)
+}
+
+func (us *UdpSession) SetReadDeadline(t time.Time) error {
+	if t == zeroTime {
+		us.readContext, us.readCancel = context.WithCancel(context.Background())
+	} else {
+		us.readContext, us.readCancel = context.WithDeadline(context.Background(), t)
+	}
+
+	conn, err := us.GetEstablishedUDPConn()
+	if err != nil {
+		return err
+	}
+	return conn.SetReadDeadline(t)
+}
+
+func (us *UdpSession) SetWriteDeadline(t time.Time) error {
+	conn, err := us.GetEstablishedUDPConn()
+	if err != nil {
+		return err
+	}
+	return conn.SetWriteDeadline(t)
+}
+
+func (us *UdpSession) SetWriteBuffer(size int) error {
+	conn, err := us.GetEstablishedUDPConn()
+	if err != nil {
+		return err
+	}
+	return conn.SetWriteBuffer(size)
+}
+
+func (us *UdpSession) SetReadBuffer(size int) error {
+	conn, err := us.GetEstablishedUDPConn()
+	if err != nil {
+		return err
+	}
+	return conn.SetReadBuffer(size)
+}
+
+func (us *UdpSession) IsClosed() bool {
+	_, err := us.GetEstablishedUDPConn()
+	if err != nil {
+		return true
+	}
+	return false
+}
+
+func (us *UdpSession) LocalAddr() net.Addr {
+	conn, err := us.GetEstablishedUDPConn()
+	if err != nil {
+		return nil
+	}
+	return conn.LocalAddr()
+}
+
+func (us *UdpSession) RemoteAddr() net.Addr {
+	if us.isListener {
+		return nil
+	}
+
+	us.RLock()
+	defer us.RUnlock()
+	if us.peer != nil {
+		return us.peer.udpAddr
+	} else {
+		return nil
+	}
+}
+
+// The endpoint close session actively.
+func (us *UdpSession) Close() (err error) {
+	us.sendCloseMsg(nil)
+
+	if conn, _ := us.GetEstablishedUDPConn(); conn != nil {

```

Comment:
Should be `err != nil` so we won't close the conn twice

Reply:
Yes, modified.